stl/inc/__msvc_string_view.hpp

```diff
@@ -221,7 +221,15 @@ struct _WChar_traits : private _Char_traits<_Elem, unsigned short> {
         // compare [_First1, _First1 + _Count) with [_First2, ...)
 #if _HAS_CXX17
         if constexpr (is_same_v<_Elem, wchar_t>) {
+#if _HAS_CXX20
+            if (_STD is_constant_evaluated()) {
+                return __builtin_wmemcmp(_First1, _First2, _Count);
+            } else {
+                return _CSTD wmemcmp(_First1, _First2, _Count);
+            }
+#else // ^^^ _HAS_CXX20 / !_HAS_CXX20 vvv
             return __builtin_wmemcmp(_First1, _First2, _Count);
+#endif // ^^^ !_HAS_CXX20 ^^^
         } else {
             return _Primary_char_traits::compare(_First1, _First2, _Count);
         }

```

Comment:
After simplifying the logic to extend the optimization to exactly C++17, another issue can be seen. The `#if _HAS_CXX17` is handling "we need to be `_CONSTEXPR17`". The `if constexpr (is_same_v<_Elem, wchar_t>)` is handling "`__builtin_wmemcmp` works for exactly `wchar_t`, otherwise we need the plain loop in `_Primary_char_traits` to be `constexpr`". But now, since you're making `wmemcmp()` more efficient, you're additionally checking `_Is_constant_evaluated()` to see whether we need to be `constexpr`. By checking *within* `is_same_v<_Elem, wchar_t>`, the optimization is being withheld from `char16_t` (and `unsigned short`, not that I care about fake `wchar_t`), and the `wmemcmp()` call is being duplicated.

We should instead rearrange the control flow to solve both problems at once.